### PR TITLE
feat(examples): Add a simple FlashAttention Implementation.

### PR DESCRIPTION
**Pull request by KuangjuX**
Draft. Don't review

### CODE REVIEW

**Comment by KuangjuX** [comment]
I want to merge this part of the code first. Although it hasn't undergone rigorous correctness verification, it's already quite substantial. I will add test code in the next PR.

File: examples/cpp/flash_attn/flash_attn.cu [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+        broadcast_mul(prev_norm_vec, rO);
+
+        broadcast_mul(cur_norm_vec, unnormized_attn_block);
+
+        block_add(rO, unnormized_attn_block, rO);
+
+        // if (tiledcuda::thread0()) {
+        //     unnormized_attn_block.dump_value();
+        //     rO.dump_value();
+        // }
```
Comment (haruhi55) [inline examples/cpp/flash_attn/flash_attn.cu:196]: At line 193, the debug print statements are commented out, while the debug print statements from lines 149 to 157 are still active. I suggest wrapping all debug print statements in a macro, as we have done before, to keep the code neat and maintain consistency.

```c
#ifdef DEBUG
...
#endif
```

File: examples/cpp/flash_attn/flash_attn.cu [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+    // run<FlashAttentionShape<64 /*M*/, 64 /*N*/, 128 /*K*/, 128 /*P*/>,
+    //     FlashAttentionShape<64 /*kTM*/, 64 /*kTN*/, 128 /*kTK*/, 128
+    //     /*kTP*/>, 2>();
+
+    // run<B2BGemmShape<1024 /*M*/, 128 /*N*/, 128 /*K*/, 128 /*P*/>,
+    //     B2BGemmShape<64 /*kTM*/, 128 /*kTN*/, 128 /*kTK*/, 128 /*kTP*/>,
+    //     2>();
+
+    return 0;
```
Comment (haruhi55) [inline examples/cpp/flash_attn/flash_attn.cu:375]: I think it is OK to delete line 344 ~ 359 and line 367~373. They can be added back when necessary in future.

**Review by haruhi55 — COMMENTED** [review]
LGTM, I like this neat codes. 😊

File: examples/cpp/flash_attn/flash_attn.cu (hunk 2 above)
Comment (KuangjuX) [reply examples/cpp/flash_attn/flash_attn.cu:346]: Deleted.

File: examples/cpp/flash_attn/flash_attn.cu (hunk 1 above)
Comment (KuangjuX) [reply examples/cpp/flash_attn/flash_attn.cu]: Fixed. And I added `ENABLE_DEBUG` option in CMake.

File: examples/cpp/flash_attn/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
+project(b2b_gemm_example LANGUAGES C CXX CUDA)
```
Comment (lcy-seso) [inline examples/cpp/flash_attn/CMakeLists.txt:2]: the project name should be changed. 😊

Comment (haruhi55) [inline examples/cpp/flash_attn/CMakeLists.txt:2]: the project name should be changed.

Comment (KuangjuX) [reply examples/cpp/flash_attn/CMakeLists.txt]: Fixed.

**Review by haruhi55 — APPROVED** [review]
LGTM

**Pull request merged**